device/cuda/src/clusterization/clusterization_algorithm.cu

```diff
@@ -23,38 +26,60 @@ namespace kernels {
     measurement_collection_types::view measurements_view,
-    vecmem::data::vector_view<unsigned int> cell_links) {
-
-    __shared__ unsigned int partition_start, partition_end;
-    __shared__ unsigned int outi;
+    vecmem::data::vector_view<unsigned int> cell_links,
+    vecmem::data::vector_view<device::details::index_t> f_backup_view,
+    vecmem::data::vector_view<device::details::index_t> gf_backup_view,
+    vecmem::data::vector_view<unsigned char> adjc_backup_view,
+    vecmem::data::vector_view<device::details::index_t> adjv_backup_view,
+    vecmem::device_atomic_ref<unsigned int> backup_mutex) {
+
+    __shared__ std::size_t partition_start, partition_end;
+    __shared__ std::size_t outi;
     extern __shared__ device::details::index_t shared_v[];
+
+    using vector_size_t =
+        vecmem::data::vector_view<device::details::index_t>::size_type;
+
     vecmem::data::vector_view<device::details::index_t> f_view{
-        max_cells_per_partition, shared_v};
+        static_cast<vector_size_t>(cfg.max_partition_size()), shared_v};
     vecmem::data::vector_view<device::details::index_t> gf_view{
-        max_cells_per_partition, shared_v + max_cells_per_partition};
+        static_cast<vector_size_t>(cfg.max_partition_size()),
+        shared_v + cfg.max_partition_size()};
     traccc::cuda::barrier barry_r;
 
-    device::ccl_kernel(threadIdx.x, blockDim.x, blockIdx.x, cells_view,
-                       modules_view, max_cells_per_partition,
-                       target_cells_per_partition, partition_start,
-                       partition_end, outi, f_view, gf_view, barry_r,
-                       measurements_view, cell_links);
+    device::ccl_kernel(cfg, threadIdx.x, blockDim.x, blockIdx.x, cells_view,
+                       modules_view, partition_start, partition_end, outi,
+                       f_view, gf_view, f_backup_view, gf_backup_view,
+                       adjc_backup_view, adjv_backup_view, backup_mutex,
+                       barry_r, measurements_view, cell_links);
 }
 
 }  // namespace kernels
 
 clusterization_algorithm::clusterization_algorithm(
     const traccc::memory_resource& mr, vecmem::copy& copy, stream& str,
-    const unsigned short target_cells_per_partition)
+    const config_type& config)
     : m_mr(mr),
       m_copy(copy),
       m_stream(str),
-      m_target_cells_per_partition(target_cells_per_partition) {}
+      m_config(config),
+      m_f_backup(m_config.backup_size(), m_mr.main),
+      m_gf_backup(m_config.backup_size(), m_mr.main),
+      m_adjc_backup(m_config.backup_size(), m_mr.main),
+      m_adjv_backup(m_config.backup_size() * 8, m_mr.main),
+      m_backup_mutex(vecmem::make_unique_alloc<unsigned int>(m_mr.main)) {
+    m_copy.get().setup(m_f_backup)->ignore();
+    m_copy.get().setup(m_gf_backup)->ignore();
+    m_copy.get().setup(m_adjc_backup)->ignore();
+    m_copy.get().setup(m_adjv_backup)->ignore();
+    TRACCC_CUDA_ERROR_CHECK(cudaMemset(
```

Comment:
This could be made into an asynchronous call, right? :thinking:

Reply:
In principle yes, but this is executed one during the entire application during the setup, so I don't think the additional complexity is worth it!

Reply:
Additional complexity? :confused: You already have the stream right there. All the `vecmem::copy::setup` calls are implicitly asynchronous themselves. Switching to [cudaMemsetAsync](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g7c9761e21d9f0999fd136c51e7b9b2a0) seems like a pretty small change here. :thinking: